File: app/src/main/java/com/zeapo/pwdstore/autofill/AutofillActivity.java
package com.zeapo.pwdstore.autofill;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.IntentSender;
import android.content.SharedPreferences;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;

import com.zeapo.pwdstore.PasswordStore;

import org.eclipse.jgit.util.StringUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// blank activity started by service for calling startIntentSenderForResult
public class AutofillActivity extends AppCompatActivity {
    public static final int REQUEST_CODE_DECRYPT_AND_VERIFY = 9913;
    public static final int REQUEST_CODE_PICK = 777;
    public static final int REQUEST_CODE_PICK_MATCH_WITH = 778;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle extras = getIntent().getExtras();

        if (extras != null && extras.containsKey("pending_intent")) {
            try {
                PendingIntent pi = extras.getParcelable("pending_intent");
                if (pi == null) {
                    return;
                }
                startIntentSenderForResult(pi.getIntentSender()
                        , REQUEST_CODE_DECRYPT_AND_VERIFY, null, 0, 0, 0);
            } catch (IntentSender.SendIntentException e) {
                Log.e(AutofillService.Constants.TAG, "SendIntentException", e);
            }
        } else if (extras != null && extras.containsKey("pick")) {
            Intent intent = new Intent(getApplicationContext(), PasswordStore.class);
            intent.putExtra("matchWith", true);
            startActivityForResult(intent, REQUEST_CODE_PICK);
        } else if (extras != null && extras.containsKey("pickMatchWith")) {
            Intent intent = new Intent(getApplicationContext(), PasswordStore.class);
            intent.putExtra("matchWith", true);
            startActivityForResult(intent, REQUEST_CODE_PICK_MATCH_WITH);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        finish();   // go back to the password field app
        switch (requestCode) {
            case REQUEST_CODE_DECRYPT_AND_VERIFY:
                if (resultCode == RESULT_OK) {
                    AutofillService.getInstance().setResultData(data);    // report the result to service
                }
                break;
            case REQUEST_CODE_PICK:
                if (resultCode == RESULT_OK) {
                    AutofillService.getInstance().setPickedPassword(data.getStringExtra("path"));
                }
                break;
            case REQUEST_CODE_PICK_MATCH_WITH:
                if (resultCode == RESULT_OK) {
                    // need to not only decrypt the picked password, but also
                    // update the "match with" preference
                    Bundle extras = getIntent().getExtras();
                    String packageName = extras.getString("packageName");
                    boolean isWeb = extras.getBoolean("isWeb");

                    String path = data.getStringExtra("path");
                    AutofillService.getInstance().setPickedPassword(data.getStringExtra("path"));

                    SharedPreferences prefs;
                    if (!isWeb) {
                        prefs = getApplicationContext().getSharedPreferences("autofill", Context.MODE_PRIVATE);
                    } else {
                        prefs = getApplicationContext().getSharedPreferences("autofill_web", Context.MODE_PRIVATE);
                    }
                    SharedPreferences.Editor editor = prefs.edit();
                    String preference = prefs.getString(packageName, "");
                    switch (preference) {
                        case "":
                        case "/first":
                        case "/never":
                            editor.putString(packageName, path);
                            break;
                        default:
                            List<String> matches = new ArrayList<>(Arrays.asList(preference.trim().split("\n")));
                            matches.add(path);
                            String paths = StringUtils.join(matches, "\n");
                            editor.putString(packageName, paths);
                    }
                    editor.apply();
                }
                break;
        }
    }
}

File: app/src/main/java/com/zeapo/pwdstore/autofill/AutofillActivity.kt
/*
 * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
@file:Suppress("Deprecation")

package com.zeapo.pwdstore.autofill

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.IntentSender
import android.content.SharedPreferences
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.edit
import com.github.ajalt.timberkt.Timber.tag
import com.github.ajalt.timberkt.e
import com.zeapo.pwdstore.PasswordStore
import com.zeapo.pwdstore.utils.splitLines
import org.eclipse.jgit.util.StringUtils

// blank activity started by service for calling startIntentSenderForResult
class AutofillActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val extras = intent.extras

        if (extras != null && extras.containsKey("pending_intent")) {
            try {
                val pi = extras.getParcelable<PendingIntent>("pending_intent") ?: return
                startIntentSenderForResult(pi.intentSender, REQUEST_CODE_DECRYPT_AND_VERIFY, null, 0, 0, 0)
            } catch (e: IntentSender.SendIntentException) {
                tag(AutofillService.Constants.TAG).e(e) { "SendIntentException" }
            }
        } else if (extras != null && extras.containsKey("pick")) {
            val intent = Intent(applicationContext, PasswordStore::class.java)
            intent.putExtra("matchWith", true)
            startActivityForResult(intent, REQUEST_CODE_PICK)
        } else if (extras != null && extras.containsKey("pickMatchWith")) {
            val intent = Intent(applicationContext, PasswordStore::class.java)
            intent.putExtra("matchWith", true)
            startActivityForResult(intent, REQUEST_CODE_PICK_MATCH_WITH)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        finish() // go back to the password field app
        when (requestCode) {
            REQUEST_CODE_DECRYPT_AND_VERIFY -> if (resultCode == RESULT_OK) {
                require(data != null)
                AutofillService.instance?.setResultData(data) // report the result to service
            }
            REQUEST_CODE_PICK -> if (resultCode == RESULT_OK) {
                require(data != null)
                AutofillService.instance?.setPickedPassword(data.getStringExtra("path")!!)
            }
            REQUEST_CODE_PICK_MATCH_WITH -> if (resultCode == RESULT_OK) {
                require(data != null)
                // need to not only decrypt the picked password, but also
                // update the "match with" preference
                val extras = intent.extras ?: return
                val packageName = extras.getString("packageName")
                val isWeb = extras.getBoolean("isWeb")

                val path = data.getStringExtra("path")
                AutofillService.instance?.setPickedPassword(data.getStringExtra("path")!!)

                val prefs: SharedPreferences
                prefs = if (!isWeb) {
                    applicationContext.getSharedPreferences("autofill", Context.MODE_PRIVATE)
                } else {
                    applicationContext.getSharedPreferences("autofill_web", Context.MODE_PRIVATE)
                }
                prefs.edit {
                    when (val preference = prefs.getString(packageName, "")) {
                        "", "/first", "/never" -> putString(packageName, path)
                        else -> {
                            val matches = arrayListOf(*preference!!.trim { it <= ' ' }.splitLines())
                            matches.add(path)
                            val paths = StringUtils.join(matches, "\n")
                            putString(packageName, paths)
                        }
                    }
                }
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    companion object {

        const val REQUEST_CODE_DECRYPT_AND_VERIFY = 9913
        const val REQUEST_CODE_PICK = 777
        const val REQUEST_CODE_PICK_MATCH_WITH = 778
    }
}
